dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,298 @@
+    DEFAULT_BATCH_SIZE = 10
+    # Default response parsing function for batch_size=1 - Simply assigns the response to the response column
+    DEFAULT_RESPONSE_PARSER = _parse_batch_response_default
+    # Default prefix to add to output columns - should be overriden for personalized output
+    DEFAULT_OUTPUT_COLUMN_PREFIX = "output"
+    # Default dictionary of output column names (key) and their descriptions (value)
+    OUTPUT_COLUMN_NAME_DESCRIPTIONS = OrderedDict(
+        [
+            ("response", "Raw response in JSON format"),
+            ("error_message", "Error message"),
+            ("error_type", "Error type or code"),
+            ("error_raw", "Raw error"),
+        ]
+    )
+    # By default, set verbose to False assuming error message and type are enough information in the logs
+    DEFAULT_VERBOSE = False
+
+    def __init__(
+        self,
+        function: Callable[[Union[Dict, List[Dict]]], Union[Dict, List[Dict]]],
+        error_handling: ErrorHandling = ErrorHandling.LOG,
+        exceptions_to_catch: Tuple[Exception] = (),
+        parallel_workers: int = DEFAULT_PARALLEL_WORKERS,
+        batch_support: bool = DEFAULT_BATCH_SUPPORT,
+        batch_size: int = DEFAULT_BATCH_SIZE,
+        batch_response_parser: Callable[[List[Dict], Any, NamedTuple], List[Dict]] = DEFAULT_RESPONSE_PARSER,
+        output_column_prefix: AnyStr = DEFAULT_OUTPUT_COLUMN_PREFIX,
+        verbose: bool = DEFAULT_VERBOSE,
+    ):
+        self.function = function
+        self.error_handling = error_handling
+        self.exceptions_to_catch = exceptions_to_catch
+        if error_handling == ErrorHandling.LOG and not exceptions_to_catch:
+            raise ValueError(
+                "Please set at least one exception in exceptions_to_catch")
+        self.parallel_workers = parallel_workers
+        self.batch_support = batch_support
+        if not batch_support:
+            # Overwrite necessary args for row-by-row iteration
+            batch_size = 1
+            batch_response_parser = _parse_batch_response_default
+            
+        self.batch_size = batch_size
+        self.batch_response_parser = batch_response_parser
+        self.output_column_prefix = output_column_prefix
+        self.verbose = verbose
+        self._output_column_names = None  # Will be set at runtime by the run method
+
+    def _get_unique_output_column_names(self, existing_names: List[AnyStr]) -> NamedTuple:
+        """Returns a named tuple with prefixed column names and their descriptions"""
+        OutputColumnNameTuple = namedtuple(
+            "OutputColumnNameTuple", self.OUTPUT_COLUMN_NAME_DESCRIPTIONS.keys())
+        return OutputColumnNameTuple(
+            *[
+                generate_unique(
+                    name=column_name, existing_names=existing_names, prefix=self.output_column_prefix)
+                for column_name in OutputColumnNameTuple._fields
+            ]
+        )
+
+    def _apply_function_and_parse_response(
```

Comment:
Having function names with `and` clearly violates the single responsibility principle.
`batch_response_parser` call can be called after the parallelization in the `post_process` function suggested bellow

Reply:
Good idea

Reply:
I thought again about this. It is not as straightforward as it looks.

This "response parsing" is responsible for error logging. If we were to move this to a post-processing step, the user would not longer be able to see the errors as they happen. All errors would be displayed at once, after the parallelization. This would have a negative impact on the usability of the feature.

The purpose of this error logging is for users to be able to monitor how the API is behaving in the job logs. If a majority of API calls are failing with the same error then the user would want to abort the job. The types of errors I am referring to are functional e.g., "this image request is too big", or "this language is not supported".  That is all the more important since most APIs make you pay for all requests, even those who fail.

I think the confusion comes from the term "response parsing". Effectively, this method is not about response parsing. It's about error logging. The fact that you need to parse the response only stems from the fact that in the batch case, the API always returns a "successful" response, even if all requests failed with functional errors.
That is why the function not only has a try/except clause to catch both technical errors (the network is down) and functional errors (you get an API response which contains an error).
I believe this subtle but important technicality was more visible before, when there were a distinct row-by-row method and a batch method. Now everything is batch, which is cleaner in terms of code, but makes these peculiarities with error handling more complex.

I would advocate for a simple solution, which is to rename the method "_apply_function_with_error_logging".

What do you think?

Happy to discuss about it over a call, as this is not a simple topic.

Reply:
Thanks for clarifying ! I also think that this code is more about error logging (even when reading the response to re throw errors).
 
So I agree with you and I think we should do the following: 
- Renaming parse response to error logging at various places in this code
- Renaming this method `_apply_function_with_error_logging`

Reply:
Yep, added in 3f3330f
I have renamed the method, and kept the references to "parsing response" only when it is clear it is required in the case of batch to parse the function results into responses and errors. Let me know what you think.